torch_xla/experimental/fori_loop.py

```diff
@@ -10,80 +10,139 @@
+  fake_iter_input = fake_carried_inputs[:-1]
+  fake_output = fake_carried_inputs[-1]
+
+  #  ====== additional_inputs_list_cond ======
+  fake_additiona_args = []
+  for additional_input in additional_inputs:
+    device = additional_input.device
+    fake_additiona_args.append(
+        torch.randint(
+            10, additional_input.size(),
+            dtype=additional_input.dtype).to(device))
+
+  #  ====== additional_inputs_list_cond ======
+  dummy_inputs_list = [
+      fake_carried_inputs[0],
+  ] + fake_additiona_args + fake_carried_inputs[1:]
+
+  if additional_inputs or fake_tensor:
+    body_fn_inputs = [
+        carried_inputs[0],
+    ] + fake_carried_inputs[1:] + list(additional_inputs)
+    cond_fn_inputs = carried_inputs + additional_inputs
   else:
-    shapes = xb.tensor_shape((fake_carried_inputs))
-  builder = xb.create_builder('test_while')
-  params = []
-  for shape in shapes:
-    p = xb.mkparam(builder, len(params), shape)
-    params.append(p)
+    body_fn_inputs = carried_inputs + additional_inputs
+    cond_fn_inputs = carried_inputs + additional_inputs
+
+  #  ====== body_fn ======
+  body_result = body_fn(*body_fn_inputs)
+  body_ctx = torch_xla._XLAC.lowering.LoweringContext()
+  body_ctx.set_name_string("bodyctx")
+
+  #  ====== body xlacomputation ======
+  body_ctx.buildforiloop(list(body_result), dummy_inputs_list)
+  body_hlo = body_ctx.hlo()
+  body_computation = xb.computation_from_module_proto("bodycomputation",
+                                                      body_hlo)
 
-  # generate cond_fn xlacomputation
-  cond_result = cond_fn(*fake_carried_inputs)
+  #  ====== cond_fn ======
+  cond_result = cond_fn(*cond_fn_inputs)
   cond_ctx = torch_xla._XLAC.lowering.LoweringContext()
   cond_ctx.set_name_string("condctx")
-  cond_ctx.buildforiloop([cond_result], list(fake_carried_inputs[2:]))
+
+  #  ====== cond xlacomputation ======
+  cond_ctx.buildforiloop([cond_result], dummy_inputs_list)
   cond_hlo = cond_ctx.hlo()
   cond_computation = xb.computation_from_module_proto("condcomputation",
                                                       cond_hlo)
 
-  # generate body_fn xlacomputation
-  body_result = body_fn(*fake_carried_inputs)
-  body_ctx = torch_xla._XLAC.lowering.LoweringContext()
-  body_ctx.set_name_string("bodyctx")
-  body_ctx.buildforiloop(list(body_result), [])
-  body_hlo = body_ctx.hlo()
-  body_computation = xb.computation_from_module_proto("bodycomputation",
-                                                      body_hlo)
+  #  ====== xla::while ======
+  iter_value = carried_inputs[0]
+  input_and_outputs_value = carried_inputs[1:]
+  total_inputs = tuple([
+      iter_value,
+  ]) + tuple(additional_inputs) + tuple(input_and_outputs_value)
+
+  kwargs = {}
+  if type(total_inputs) is tuple:
+    shapes = xb.tensor_shape(total_inputs)
+  else:
+    shapes = xb.tensor_shape((total_inputs))
+  builder = xb.create_builder('test_while')
```

Comment:
why is it called `test_while`?

Reply:
thanks for catch, updated to `while_loop`